src/sys/arm_compat/syscall/mod.rs

```diff
@@ -2,88 +2,136 @@
-    SYS_FLEN = 0x0C,
+    SYS_FLEN,
     // /// [SYS_TMPNAM (0x0D)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#621sys_tmpnam-0x0d)
     // #[deprecated = "tmpnam is deprecated as not secure on most host systems"]
     //  SYS_TMPNAM = 0x0D,
     /// [SYS_REMOVE (0x0E)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#615sys_remove-0x0e)
-    SYS_REMOVE = 0x0E,
+    SYS_REMOVE,
     /// [SYS_RENAME (0x0F)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#616sys_rename-0x0f)
-    SYS_RENAME = 0x0F,
+    SYS_RENAME,
     /// [SYS_CLOCK (0x10)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#sys-clock-0x10)
-    SYS_CLOCK = 0x10,
+    SYS_CLOCK,
     /// [SYS_TIME (0x11)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#620sys_time-0x11)
-    SYS_TIME = 0x11,
+    SYS_TIME,
     /// [SYS_SYSTEM (0x12)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#618sys_system-0x12)
-    SYS_SYSTEM = 0x12,
+    SYS_SYSTEM,
     /// [SYS_ERRNO (0x13)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#64sys_errno-0x13)
-    SYS_ERRNO = 0x13,
+    SYS_ERRNO,
     /// [SYS_GET_CMDLINE (0x15)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#68sys_get_cmdline-0x15)
-    SYS_GET_CMDLINE = 0x15,
+    SYS_GET_CMDLINE,
     /// [SYS_HEAPINFO (0x16)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#69sys_heapinfo-0x16)
-    SYS_HEAPINFO = 0x16,
+    SYS_HEAPINFO,
     // #[deprecated = "obsoleted in semihosting specification version 2.0"]
     //  angel_SWIreason_EnterSVC = 0x17,
     /// [SYS_EXIT (0x18)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#65sys_exit-0x18)
     #[doc(alias = "angel_SWIreason_ReportException")] // old name
-    SYS_EXIT = 0x18,
+    SYS_EXIT,
     // #[deprecated = "obsoleted in semihosting specification version 2.0"]
     //  angelSWI_Reason_SyncCacheRange = 0x19,
     /// [SYS_EXIT_EXTENDED (0x20)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#66sys_exit_extended-0x20)
     // we use SYS_EXIT on 64-bit system because SYS_EXIT_EXTENDED is identical to the behavior of the mandatory SYS_EXIT.
     #[cfg_attr(target_pointer_width = "64", allow(dead_code))]
-    SYS_EXIT_EXTENDED = 0x20,
+    SYS_EXIT_EXTENDED,
     /// [SYS_ELAPSED (0x30)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#63sys_elapsed-0x30)
-    SYS_ELAPSED = 0x30,
+    SYS_ELAPSED,
     /// [SYS_TICKFREQ (0x31)](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#619sys_tickfreq-0x31)
-    SYS_TICKFREQ = 0x31,
+    SYS_TICKFREQ,
+
+    /// Some other (e.g. user defined) operation number
+    Other(u32),
```

Comment:
I'm not sure if it is really a good idea to allow arbitrary numbers. Since the range of user-defined syscalls is clearly defined, it might be better to add a constructor that only accepts numbers that fall within that range, so that it is less likely to be misused.

Something like:

```rs
pub struct OperationNumber(u32);

impl OperationNumber {
    const SYS_OPEN: Self = Self(0x01);
    // ...

    #[inline] // inline to help compiler to remove assertion
    pub const fn user_defined(number: u32) -> Self {
        assert!(number >= 0x100 && number <= 0x1FF);
        Self(number)
    }
}
```

Reply:
Good point. This is a lot cleaner.

I could not find a user-defined range of operation codes for MIPS though... How do you want to proceed there?